### PR TITLE
Call out fields for tower_credentials

### LINKED ISSUE #32352: tower_credential module unable to POST correctly
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
`tower_credential`

##### ANSIBLE VERSION
```
ansible 2.4.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/lmadsen/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.13 (default, Sep  5 2017, 08:53:59) [GCC 7.1.1 20170622 (Red Hat 7.1.1-3)]
```

##### CONFIGURATION
```
DEFAULT_HOST_LIST(/home/lmadsen/src/github/leifmadsen/awx-builder/ansible.cfg) = [u'/home/lmadsen/src/github/leifmadsen/awx-builder/inventory']
HOST_KEY_CHECKING(/home/lmadsen/src/github/leifmadsen/awx-builder/ansible.cfg) = False
RETRY_FILES_ENABLED(/home/lmadsen/src/github/leifmadsen/awx-builder/ansible.cfg) = False
```

##### OS / ENVIRONMENT
Fedora 26

##### SUMMARY
Use of the `tower_credential` module fails to post to the API

##### STEPS TO REPRODUCE
Run the following play against AWX and you'll be returned the following failure:

Sample play
```
    - name: Configure SCM Credential
      tower_credential:
        name: Git
        description: Git default credential
        organization: null
        kind: scm
        tower_verify_ssl: False
```

##### EXPECTED RESULTS
No failure.

##### ACTUAL RESULTS
```
The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_GcQSWf/ansible_module_tower_credential.py", line 292, in <module>
    main()
  File "/tmp/ansible_GcQSWf/ansible_module_tower_credential.py", line 277, in main
    result = credential.modify(**params)
  File "/home/lmadsen/.local/lib/python2.7/site-packages/tower_cli/models/base.py", line 961, in modify
    force_on_exists=True, **kwargs)
  File "/home/lmadsen/.local/lib/python2.7/site-packages/tower_cli/models/base.py", line 345, in write
    r = getattr(client, method.lower())(url, data=kwargs)
  File "/home/lmadsen/.local/lib/python2.7/site-packages/requests/sessions.py", line 577, in patch
    return self.request('PATCH', url, data=data, **kwargs)
  File "/home/lmadsen/.local/lib/python2.7/site-packages/tower_cli/api.py", line 235, in request
    kwargs.get('data', None), r.content.decode('utf8'))
tower_cli.exceptions.BadRequest: The Tower server claims it was sent a bad request.

PATCH http://awx.nfvpe.site/api/v1/credentials/6/
Params: None
Data: {"authorize": false, "state": "present", "description": "Git default credential", "kind": "scm", "name": "Git", "tower_verify_ssl": false}

Response: {"detail":["'state' is not a valid field for Source Control"]}

fatal: [127.0.0.1]: FAILED! => {
    "changed": false, 
    "failed": true, 
    "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_GcQSWf/ansible_module_tower_credential.py\", line 292, in <module>\n    main()\n  File \"/tmp/ansible_GcQSWf/ansible_module_tower_credential.py\", line 277, in main\n    result = credential.modify(**params)\n  File \"/home/lmadsen/.local/lib/python2.7/site-packages/tower_cli/models/base.py\", line 961, in modify\n    force_on_exists=True, **kwargs)\n  File \"/home/lmadsen/.local/lib/python2.7/site-packages/tower_cli/models/base.py\", line 345, in write\n    r = getattr(client, method.lower())(url, data=kwargs)\n  File \"/home/lmadsen/.local/lib/python2.7/site-packages/requests/sessions.py\", line 577, in patch\n    return self.request('PATCH', url, data=data, **kwargs)\n  File \"/home/lmadsen/.local/lib/python2.7/site-packages/tower_cli/api.py\", line 235, in request\n    kwargs.get('data', None), r.content.decode('utf8'))\ntower_cli.exceptions.BadRequest: The Tower server claims it was sent a bad request.\n\nPATCH http://awx.nfvpe.site/api/v1/credentials/6/\nParams: None\nData: {\"authorize\": false, \"state\": \"present\", \"description\": \"Git default credential\", \"kind\": \"scm\", \"name\": \"Git\", \"tower_verify_ssl\": false}\n\nResponse: {\"detail\":[\"'state' is not a valid field for Source Control\"]}\n", 
    "module_stdout": "", 
    "msg": "MODULE FAILURE", 
    "rc": 0
}
```

### PR DESCRIPTION
**Pull request by leifmadsen**
##### SUMMARY
When passing **params directly to credential.modify(), you get a rejection
by the API due to fields that it doesn't accept (such as 'state'). Call
out the individual fields like in tower_project.py results in a successful
call to the v1 API.

Tested both state: present, and state: absent. Also fixed two
small documentation typos.

Fixes #32352

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
`tower_credential`

##### ANSIBLE VERSION
```
ansible 2.5.0 (bug/issue_32352_tower_credentials 7c68bb0c1a) last updated 2017/10/31 09:56:20 (GMT -400)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/lmadsen/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/lmadsen/src/github/leifmadsen/ansible/lib/ansible
  executable location = /home/lmadsen/src/github/leifmadsen/ansible/bin/ansible
  python version = 2.7.13 (default, Sep  5 2017, 08:53:59) [GCC 7.1.1 20170622 (Red Hat 7.1.1-3)]
```

##### ADDITIONAL INFORMATION
Working playbook
```
# vim: ft=yaml tabstop=2 shiftwidth=2 expandtab
---

- hosts: localhost
  pre_tasks:
    - name: Check for tower_cli configuration
      stat:
        path: "{{ ansible_env.HOME }}/.tower_cli.cfg"
      register: tower_cli_cfg

    - name: Fail if tower_cli configuration doesn't exist
      fail:
        msg: "You need to configure Tower CLI in ~/.tower_cli.cfg"
      when: not tower_cli_cfg.stat.exists

  tasks:
    - name: Configure SCM Credential
      tower_credential:
        name: Git
        description: Git default credential
        organization: Default
        kind: scm
        state: present
        tower_verify_ssl: False

    - name: Configure projects
      tower_project:
        name: "openstack-builder-testing"
        organization: Default
        state: present
        scm_type: git
        scm_url: "https://github.com/leifmadsen/openstack-inventory-builder.git"
        scm_branch: master
        scm_clean: yes
        scm_credential: Git
        scm_delete_on_update: yes
        scm_update_on_launch: yes
        tower_verify_ssl: False
```

`state: absent`
```
ansible-playbook -i ~/src/github/leifmadsen/awx-builder/inventory/localhost ~/src/github/leifmadsen/awx-builder/site.yml 

PLAY [localhost] ******************************************************************************************************************************************************************************************************************************

TASK [Gathering Facts] ************************************************************************************************************************************************************************************************************************
ok: [127.0.0.1]

TASK [Check for tower_cli configuration] ******************************************************************************************************************************************************************************************************
ok: [127.0.0.1]

TASK [Fail if tower_cli configuration doesn't exist] ******************************************************************************************************************************************************************************************
skipping: [127.0.0.1]

TASK [Configure SCM Credential] ***************************************************************************************************************************************************************************************************************
changed: [127.0.0.1]

TASK [Configure projects] *********************************************************************************************************************************************************************************************************************
changed: [127.0.0.1]

PLAY RECAP ************************************************************************************************************************************************************************************************************************************
127.0.0.1                  : ok=4    changed=2    unreachable=0    failed=0   

```

`state: present`
```
ansible-playbook -i ~/src/github/leifmadsen/awx-builder/inventory/localhost ~/src/github/leifmadsen/awx-builder/site.yml 

PLAY [localhost] ******************************************************************************************************************************************************************************************************************************

TASK [Gathering Facts] ************************************************************************************************************************************************************************************************************************
ok: [127.0.0.1]

TASK [Check for tower_cli configuration] ******************************************************************************************************************************************************************************************************
ok: [127.0.0.1]

TASK [Fail if tower_cli configuration doesn't exist] ******************************************************************************************************************************************************************************************
skipping: [127.0.0.1]

TASK [Configure SCM Credential] ***************************************************************************************************************************************************************************************************************
changed: [127.0.0.1]

TASK [Configure projects] *********************************************************************************************************************************************************************************************************************
changed: [127.0.0.1]

PLAY RECAP ************************************************************************************************************************************************************************************************************************************
127.0.0.1                  : ok=4    changed=2    unreachable=0    failed=0 
```

### CODE REVIEW

**Comment by leifmadsen** [comment]
Request for this to be backported to stable-2.4 as well please.

**Comment by leifmadsen** [comment]
I changed the approach to this to be a lot simpler. I only needed to drop the `state` value from the dictionary, so used `.pop()` instead of building out a complicated list that we'd have to keep up to date.

**Comment by challett** [comment]
I've had issues with `tower_host` being passed in the credentials module as well with tower-cli 3.1.8 (since ansible modules do not support 3.2.0 as mentioned in #31655).  Any way that we can modify this PR to support removing that parameter as well? The other tower_* modules do not get complaints from tower when tower_host is supplied.

The result is the same when using `tower_config_file`.  I imagine that tower would also complain about `tower_username` and `tower_password`

This may actually be a tower-cli issue on second thought.

**Comment by leifmadsen** [comment]
I actually just upgraded to using tower-cli 3.2.0, and things still seem to work for me. I'll let you know how it goes.

This was the minimum amount of changes that I required for this module to work. I'm thinking of going back and trying to clean things up so all the tower modules work in the same type of way (instead of passing a list of fields).

**Comment by leifmadsen** [comment]
Actually, that's not true. It seems to work ok, until you go and do the initial creation...

**Review by wwitzel3 — APPROVED** [review]

**Comment by wwitzel3** [comment]
shipit

This will work against 3.1.8

**Comment by challett** [comment]
@leifmadsen Try passing your credentials into the module as a parameter as it is documented instead of as a config file.  With your inital changes this would have been fixed actually because the tower_* config params would have been removed. Not saying that it's the best approach but it would have solved my issue :).

Also, creating a credential with 3.2.0 complains about `Missing required fields: credential-type` as this module passes `kind`.  Passing `credential-type` instead of `kind` causes the module to complain (as expected as that's not the parameter it expects).

**Comment by leifmadsen** [comment]
@challett yea, I've been looking at how much work it would be to migrate the modules to 3.2.0, and it seems to be a lot of work. I initially started with `tower_credentials.py` and a bit of blurb like this (after I updated the use of `kind` with `credential_type`:

```
@@ -205,6 +202,7 @@ def main():
     name = module.params.get('name')
     organization = module.params.get('organization')
     state = module.params.get('state')
+    credential_type = module.params.get('credential_type')
 
     json_output = {'credential': name, 'state': state}
 
@@ -216,6 +214,10 @@ def main():
             params = module.params.copy()
             params['create_on_missing'] = True
 
+            cred_type_res = tower_cli.get_resource('credential_type')
+            cred_type = cred_type_res.get(name=credential_type)
+            params['credential_type'] = cred_type['id']
```

But the issue is that you need to actually construct the credential JSON a bit differently, because they are a lot more dynamic (not static like in the old API), so you need to look up the fields your `credential` wants for that `credential-type`, and construct an `inputs` field dynamically, passing data to it in that field.

Anyways, I kind of know what needs doing, but I'm not sure I can justify the time doing all this work when it's not really part of my day job...

File: lib/ansible/modules/web_infrastructure/ansible_tower/tower_credential.py [hunk 1]
```diff
@@ -231,6 +231,7 @@ def main():
                     params['ssh_key_data'] = f.read()
 
             if state == 'present':
+                params.pop('state')
```
Comment (AlanCoding) [inline lib/ansible/modules/web_infrastructure/ansible_tower/tower_credential.py:234]: would it be safer to do `.pop('state', None)`? I feel skeptical that the source of the parameter will consistently be there.

**Comment by ryanpetrello** [comment]
Thanks for contributing this, @leifmadsen.  I've recently added support for integration tests for the various Tower modules:

https://github.com/ansible/ansible/pull/37421

Would you mind adding similar tests for this bug fix so we can prove it works (and to avoid regressions)?  Thanks!

needs_revision

**Comment by AlanCoding** [comment]
Changes here were covered by other PRs.

close_me

**Pull request closed without merging**